src/Symfony/Component/AstGenerator/AstGeneratorInterface.php

```diff
@@ -0,0 +1,39 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\AstGenerator;
+
+/**
+ * An AstGeneratorInterface is a contract to transform an object into an AST.
+ *
+ * @author Joel Wurtz <joel.wurtz@gmail.com>
+ */
+interface AstGeneratorInterface
+{
+    /**
+     * Generate an object into an AST given a specific context.
+     *
+     * @param mixed $object  Object to generate AST from
+     * @param array $context Context for the generator
+     *
+     * @return \PhpParser\Node[] An array of statements (AST Node)
+     */
+    public function generate($object, array $context = array());
+
+    /**
+     * Check whether the given object is supported for generation by this generator.
+     *
+     * @param mixed $object Object to generate AST from
+     *
+     * @return bool
+     */
+    public function supportsGeneration($object);
```

Comment:
> And replacing a top level routing generator by the other would not be useful indeed but you could chain them to put them in the same file for example.

Still, I don't see why `UrlGeneratorGenerator` and  `PhpMatcherGenerator` should implement the generic `generate($object, array $context)`:
```php
final class UrlGeneratorAstGenerator 
{
    public function generate(
        RouteCollection $routeCollection,
        string $urlGeneratorClass,
        string $baseUrlGeneratorClass
    ): NodeList 
    {
        // ...
    }
}

final class PhpMatcherAstGenerator
{
    public function generate(RouteCollection $collection): NodeList
    {
        // ...
    }
}

$nodeList = 
    $this->urlAstGenerator->generate(...)
    ->append(
        $this->phpMatcherAstGenerator->generate(...)
    );
```
Generic interface for them adds false flexibility.